apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,174 @@
+import { User, requestLogToServer } from "@pcd/passport-interface";
+import { PCDCollection } from "@pcd/pcd-collection";
+import {
+  SemaphoreIdentityPCD,
+  SemaphoreIdentityPCDPackage
+} from "@pcd/semaphore-identity-pcd";
+import { appConfig } from "./appConfig";
+import { AppState } from "./state";
+
+/**
+ * Determines whether the app's global state as represented by {@link AppState} object
+ * contains valid data. If it does not, returns the set of things that are incorrect about
+ * it in an array of human interpretable strings. If there are no errors, returns an
+ * empty array.
+ */
+export function validateState(state: AppState): string[] {
+  const validationErrors: string[] = [];
+
+  if (!state.self) {
+    validationErrors.push("missing 'self' field from app state");
+  }
+
+  if (!state.identity) {
+    validationErrors.push("missing 'identity' field from app state");
+  }
+
+  if (!state.encryptionKey) {
+    validationErrors.push("missing 'encryption' field from app state key");
+  }
+
+  if (!state.pcds) {
+    validationErrors.push("missing 'pcds' field from app state");
+  }
+
+  if (state.pcds.size() === 0) {
+    validationErrors.push("'pcds' field in app state contains no pcds");
+  }
+
+  const identityPCDFromCollection = state.pcds.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )[0] as SemaphoreIdentityPCD | undefined;
+
+  if (!identityPCDFromCollection) {
+    validationErrors.push(
+      "'pcds' field in app state does not contain an identity PCD"
+    );
+  }
+
+  const identityFromPCDCollection = identityPCDFromCollection?.claim?.identity;
+  const commitmentOfIdentityPCDInCollection =
+    identityFromPCDCollection?.commitment?.toString();
+  const commitmentFromSelfField = state?.self?.commitment;
+  const commitmentFromIdentityField = state?.identity?.commitment?.toString();
+
+  if (commitmentOfIdentityPCDInCollection !== commitmentFromSelfField) {
+    validationErrors.push(
+      `commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})` +
+        ` does not match commitment in 'self' field of app state (${commitmentFromSelfField})`
+    );
+  }
+
+  if (commitmentFromSelfField !== commitmentFromIdentityField) {
+    validationErrors.push(
+      `commitment in 'self' field of app state (${commitmentFromSelfField})` +
+        ` does not match commitment of 'identity' field of app state (${commitmentFromIdentityField})`
+    );
+  }
+
+  if (commitmentFromIdentityField !== commitmentOfIdentityPCDInCollection) {
+    validationErrors.push(
+      `commitment of 'identity' field of app state (${commitmentFromIdentityField})` +
+        ` does not match commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})`
+    );
+  }
+
+  return validationErrors;
+}
+
+/**
```

Comment:
I think it's worth refactoring this code to share common parts, so that each given invariant is checked consistently in only one place.  Otherwise I predict headaches when we find out that each of these functions are checking different versions of the same invariant.

Reply:
Furthermore I think it would be worth considering passing around extra arguments in some or all cases to ensure the full suite of invariants can be checked.  Most specifically passing some form of user identity (maybe `self`, or `identity`, or maybe just the `commitment` part extracting from them) so that all cases of PCD validation can check that the user's identity PCD matches.

Reply:
> I think it's worth refactoring this code to share common parts, so that each given invariant is checked consistently in only one place. Otherwise I predict headaches when we find out that each of these functions are checking different versions of the same invariant.

I thought about this, and thought that the headache of coming up with an abstraction for this may not be worth it. For example, we invoke this validation function when we save a `PCDCollection` to localstorage inside of the `localstorage.ts` file. Unless I refactor that call site code so that it's aware of the entirety of AppState, or move the validation logic to the depth where it has access to appState and use the validation logic to determine whether to save to localstorage (which would be less than ideal imo because then we would have to remember to invoke the validation logic every time), it's not going to be easy to come up with a single function that can be called both there and in places where we do have an `AppState`. 

One way to create such a function would be to also make the `AppState` argument optional, in which case that validation function would check only the self-consistency of the passed-in values. That could get complicated.

I also thought about passing extra arguments like user or identity into this function. I'm not sure under which conditions you were thinking of passing them arround - the cases I was thinking about involved passing them in the case that the application is attempting to make an update to `AppState`, so that we could check if the resulting update would result in a consistent state. I'm not sure that I can come up with another way to cleanly check state updates prior to making them - maybe you have a suggestion?

Reply:
I agree that checking the full AppState is probably complicated enough to need to be separate.  In particular, invariants about what a valid logged-off state looks like, and which fields should always be defined in a logged-in state are unique to the initial load scenario.

For the other scenarios, the validators should only ever see a logged-in state.  Furthermore the relevant fields are just `self`, `pcds`, and `subscriptions`, which are expected to always be defined when logged in.  My idea (not fully vetted) was to have a single validator function which takes those 3 arguments, but which makes them all optional.  In this case `undefined` means "I don't have access to that data right now", and the validator checks as many invariants as it can check with the data it's given.  E.g. if given `pcds` without `self` it'll check that the identity PCD exists, but if also given `self` it can check that the commitment matches.

The only thing that approach can't validate is whether a field is illegally being set to `undefined`, but that's basically never allowed (except at logoff time) so should be relatively easy to check elsewhere (ideally in the type system once strict mode is turned on).

Reply:
I think I've resolved this.